src/codegen/llvm/codegen_x86_64.cc

```diff
@@ -0,0 +1,68 @@
+/*!
+ *  Copyright (c) 2017 by Contributors
+ * \file codegen_arm.cc
+ * \brief X86-64 specific code generator
+ */
+#ifdef TVM_LLVM_VERSION
+#include "codegen_cpu.h"
+
+namespace tvm {
+namespace codegen {
+
+class CodeGenX86_64 final : public CodeGenCPU {
+ public:
+  llvm::Value* VisitExpr_(const Cast* op) override;
+};
+
+llvm::Value* CodeGenX86_64::VisitExpr_(const Cast* op) {
+  // LLVM does not automatically generate the correct instruction sequences for
+  // half -> float conversion (using AVX2/AVX512 variants of vcvtph2ps).
+  const auto from = op->value.type();
+  const auto to = op->type;
+  if (from.is_float() && to.is_float() && from.bits() == 16 && to.bits() == 32) {
+    CHECK_EQ(from.lanes(), to.lanes());
+    CHECK_NOTNULL(target_machine_);
+
+    const auto has_f16c =
+        target_machine_->getTargetFeatureString().find("f16c") != llvm::StringRef::npos;
+    const auto has_avx512f =
+        target_machine_->getTargetFeatureString().find("avx512f") != llvm::StringRef::npos;
+
+    // TODO(tulloch): implement version generic over lanes.
+    if (from.lanes() == 8 && (has_f16c || has_avx512f)) {
+      Array<Expr> vcvt_args;
+      ::llvm::Intrinsic::ID vcvtph2ps_id = ::llvm::Intrinsic::x86_vcvtph2ps_256;
+      vcvt_args.push_back(ir::UIntImm::make(UInt(32), vcvtph2ps_id));
+      vcvt_args.push_back(ir::UIntImm::make(UInt(32), 0));
+      vcvt_args.push_back(
+          ir::Call::make(Int(16, 8), ir::Call::reinterpret, {op->value}, ir::Call::PureIntrinsic));
+      return MakeValue(ir::Call::make(to, "llvm_intrin", vcvt_args, ir::Call::PureIntrinsic));
+    }
+
+    // TODO(tulloch): implement version generic over lanes.
+    if (from.lanes() == 16 && has_avx512f) {
```

Comment:
Does `has_avx512f==true` imply `has_f16c==true`? And, is there a case that `lanes==16` but `has_avx512f==false`? I am not familiar with the F16C instruction set.